core/src/main/scala/filodb.core/memstore/PartKeyIndex.scala

```diff
@@ -205,3 +419,209 @@ trait PartKeyIndexDownsampled extends PartKeyIndexRaw {
+ * * MatchAll queries (match all values)
+ * * Range queries (start to end on a long value)
+ *
+ * This list may expand over time.
+ *
+ * The top level is always a Boolean query, even if it only has one child query.  A
+ * Boolean query will never have zero child queries.  Leaf nodes are always non-Boolean
+ * queries.
+ *
+ * For example:
+ *   + represents MUST, - represents MUST_NOT
+ *   Input query -> ((+Equals(Col, A) -Equals(Col, B)) +Regex(Col2, C.*))
+ *
+ *   The visitor call pattern would be:
+ *    * visitStartBooleanQuery
+ *    * visitStartBooleanQuery
+ *    * visitEqualsQuery(Col, A, MUST)
+ *    * visitEqualsQuery(Col, B, MUST_NOT)
+ *    * visitEndBooleanQuery
+ *    * visitRegexQuery(Col2, C.*, MUST)
+ *    * visitEndBooleanQuery
+ *
+ * This class should be used instead of putting query parsing code in each
+ * implementation so that common optimizations, such as query rewriting,
+ * can occur in one place.
+ */
+abstract class PartKeyQueryBuilder {
+  /**
+   * Start a new boolean query
+   */
+  protected def visitStartBooleanQuery(): Unit
+
+  /**
+   * Indicate the current boolean query has ended
+   */
+  protected def visitEndBooleanQuery(): Unit
+
+  /**
+   * Add a new equals query to the current boolean query
+   */
+  protected def visitEqualsQuery(column: String, term: String, occur: PartKeyQueryOccur): Unit
+
+  /**
+   * Add a new Regex query to the current boolean query
+   */
+  protected def visitRegexQuery(column: String, pattern: String, occur: PartKeyQueryOccur): Unit
+
+  /**
+   * Add a TermsIn query to the current boolean query
+   */
+  protected def visitTermInQuery(column: String, terms: Seq[String], occur: PartKeyQueryOccur): Unit
+
+  /**
+   * Add a prefix query to the current boolean query
+   */
+  protected def visitPrefixQuery(column: String, prefix: String, occur: PartKeyQueryOccur): Unit
+
+  /**
+   * Add a match all query to the current boolean query
+   */
+  protected def visitMatchAllQuery(): Unit
+
+  /**
+   * Add a range match to the current boolean query
+   */
+  protected def visitRangeQuery(column: String, start: Long, end: Long, occur: PartKeyQueryOccur): Unit
+
+  protected def visitQuery(columnFilters: Seq[ColumnFilter]): Unit = {
+    visitStartBooleanQuery()
+    columnFilters.foreach { filter =>
+      visitFilter(filter.column, filter.filter)
+    }
+    visitEndBooleanQuery()
+  }
+
+  protected def visitQueryWithStartAndEnd(columnFilters: Seq[ColumnFilter], startTime: PartitionKey,
+                                          endTime: PartitionKey): Unit = {
+    visitStartBooleanQuery()
+    columnFilters.foreach { filter =>
+      visitFilter(filter.column, filter.filter)
+    }
+    visitRangeQuery(START_TIME, 0, endTime, OccurMust)
+    visitRangeQuery(END_TIME, startTime, Long.MaxValue, OccurMust)
+    visitEndBooleanQuery()
+  }
+
+  // scalastyle:off method.length
+  private def visitFilter(column: String, filter: Filter): Unit = {
```

Comment:
This method is the existing Lucene `leafFilter` method but with visitor callbacks instead of inline object creation.